Fix rabin2 -Pj and idpij (json PDB output format) ##bin

emuresan opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

The information packed in the `pdb` field of the json, such as the `types` and `gvars` should be objects inside an array but, with recent changes in the json/pdb area, those were ommited.
The result was an invalid json when calling, for example `rabin2 -Pj some.pdb > some.pdb.json` or `idpij` in the radare shell.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

On the master branch, the issue can be reproduced as follows:
```
$ rabin2 -Pj ntdll.pdb | jq . 1>/dev/null
parse error: ':' not as part of an object at line 1, column 15
```
Also, the issue can easily be seen by inspecting the first items in the resulting json ( the missing `[{` before the `"types"` array):
```
{"pdb":"types":[{"type":"structure","name"
```

With the following fix, the resulting json is valid:
```
$ rabin2 -Pj ntdll.pdb | jq . 1>/dev/null ; echo $?
0
```
Also, it can be seen in the json as well:
```
{"pdb":[{"types":[{"type":"structure",
```

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

trufae commented:
PD. i have fixed the PR title

emuresan commented:
> Rabin2 edit ruins the json validity, but the json is valid for `idpij`, your PR fixes the Rabin2, but creates invalid json for idpij

The issue is a little bit tricky since `r_main_rabin2` ,via the `run_action` macro, expects a valid json from `r_core_pdb_info` ( with heading and trailing brackets ), while `cmd_info` expects an invalid json ( without heading and trailing brackets ). That being said, it's harder than it seems to fix one without breaking the other.

XVilka commented:
I suggest just remove "pdb" part

emuresan commented:
> I suggest just remove "pdb" part

That will fix the apparent issue, but there's better ways to do it imho.
For one, "removing the \"pdb\" part" sugests neglecting the existence of the `run_action` interface in `r_main_rabin2` which will lead to further inconsistency.
And again, having an api that has a json output and intentionally expecting that json to be invalid seems wrong.
Fixing the issue in this manner seems like we're simply paving the way for future bugs.

trufae reviewed (approved):
lgtm, we can further improve this later in a separate pr

ret2libc commented:
Hi @emuresan , is there any news on this? I see there were some doubts around the expected output. I think a test in this case would help a lot for avoiding future issues, for making sure your PR works as intended and for ensuring what is intended it is actually what everybody expects as well :) I think you can easily add a new test in `test/db/tools/rabin2`. Let us know if you need anything.

trufae commented:
Some tests are not passing

emuresan commented:
Yeah, the tests I added failed because they are relying on the `minimal.pdb` file that I added here: https://github.com/radareorg/radare2-testbins/pull/50 . A rerun of the tests should hopefully have them passing.

ret2libc commented:
Hi @emuresan , as @HoundThe I have some doubts about the resulting json with this change. Is there a reason to put things in an array?

I'm no windows/PDB expert at all, so I may be missing some important details, but I would expect something like:
```
> idpij ./test.pdb
{
   "types": [.....],
   "gvars": [....],
}
```
and for the rabin2 case something like:
```
$ rabin2 -Pj ./test.pdb
{
   "pdb": {
      "types": [....],
      "gvars": [....],
   }
}
```
Does this make sense?

Basically, I would expect we keep the object of idpij and we wrap it around "pdb", which I guess was the original intent of the code.

emuresan commented:
Well, the original code had all of those in arrays (I'm talking about the ages old build that I had on my machine before I upraded to latest master and spotted the bug).
The reason why it also makes sense to have all the information in array entries is because of the structure of the pdb files. Basically the file is grouped in streams and each stream will hold different data: https://llvm.org/docs/PDB/index.html .
Now, we don't necessarly care about said streams or data in the json output since only care about types and gvars so I can change it if you'd like.

ret2libc commented:
@HoundThe do you have an opinion on this? array vs object? In any case, however, @emuresan a valid JSON is better than an invalid one :D So whatever we choose it must be valid. Right not, it should be very easy to make it valid json in all cases (rabin2 and r2) by keeping the object json structure.

trufae commented:
@ret2libc i think this is good to merge

ret2libc commented:
> @ret2libc i think this is good to merge

I'm not sure actually.. I still don't understand why the structure is:
```json
"pdb": [
   {
      "types": [ ... ]
   },
   {
      "gvars": [ ... ]
   }
]
```
instead of:
```json
"pdb": {
      "types": [ ... ]
      "gvars": [ ... ]
},
```

trufae commented:
uhm yeah. that shouldnt be like this

trufae commented:
ping @emuresan , can you change the code to construct the proposed structure in https://github.com/radareorg/radare2/pull/17396#issuecomment-698172821

trufae commented on test/db/tools/rabin2:
```diff
@@ -1045,6 +1045,15 @@ EXPECT=<<EOF
 
+NAME=rabin2 -Pj
+FILE=bins/pdb/minimal.pdb
+CMDS=!rabin2 -Pj ${R2_FILE}
```
append `~j` to get the JSON indent output, its easier to read and check the diff in case the output changes.

trufae reviewed (changes requested):
Please see @ret2libc comment. just removing the pj_o pj_end calls will fix the issue i guess. in which situations is the output incorrect without those calls?

trufae commented:
Ping
